### PR TITLE
feat(meta, frontend): support drop cascade for tables, mviews, views, indexes, sources and sinks

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Resolve https://github.com/risingwavelabs/risingwave/issues/11245 https://github.com/risingwavelabs/risingwave/issues/7577
- Use `drop_relation` to replace `drop_table`, `drop_source`, `drop_index`, `drop_sink` and `drop_source` in meta.
- To support drop cascade, we need to analyze the dependencies between relations. Because relations can depend on each other, we need to drop them in a unified way. We use the breadth-first search method to find dependencies. Once we find out all dependencies, we handle them as usual (e.g. remove them from the etcd. notify frontend. drop streaming job if necessary).
- Some redundant code between different drop methods could also be reduced in this PR.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- SQL commands, functions, and operators

### Release note

- Support drop table, index, view, materialized view, source and sink cascade.

</details>

### CODE REVIEW

File: src/meta/src/rpc/service/ddl_service.rs [hunk 1]
```diff
@@ -241,11 +247,19 @@ where
...
     ) -> Result<Response<DropSinkResponse>, Status> {
-        let sink_id = request.into_inner().sink_id;
-
+        let request = request.into_inner();
+        let sink_id = request.sink_id;
+        let drop_mode = if request.cascade {
+            DropMode::Cascade
+        } else {
+            DropMode::Restrict
+        };
```
Comment (kwannoel) [inline src/meta/src/rpc/service/ddl_service.rs:256]: Saw it being used in a few areas, can it be abstracted to:
`let drop_mode = DropMode::from_request_setting(request.cascade)`

File: src/meta/src/stream/stream_manager.rs [hunk 2]
```diff
@@ -864,7 +867,11 @@ mod tests {
         async fn drop_materialized_views(&self, table_ids: Vec<TableId>) -> MetaResult<()> {
             for table_id in &table_ids {
                 self.catalog_manager
-                    .drop_table(table_id.table_id, vec![], self.fragment_manager.clone())
+                    .drop_relation(
+                        RelationIdEnum::Table(table_id.table_id),
+                        self.fragment_manager.clone(),
+                        DropMode::Restrict,
```
Comment (kwannoel) [inline src/meta/src/stream/stream_manager.rs:873]: Should this be inheriting `DropMode` setting from the caller?

File: src/meta/src/manager/catalog/mod.rs [hunk 3]
```diff
@@ -823,271 +791,517 @@ where
...
+        let mut deque = VecDeque::new();
+
+        // Relation dependencies is a DAG rather than a tree, so we need to use `HashSet` instead of
+        // `Vec` to record ids.        Sink
+        //          |
+        //        MView
+        //        /   \
+        //       View  |
+        //        \   /
+        //        Table
```
Comment (kwannoel) [inline src/meta/src/manager/catalog/mod.rs:819]:
```suggestion
        // `Vec` to record ids.
        //        Sink
        //          |
        //        MView
        //        /   \
        //       View  |
        //        \   /
        //        Table
```

File: src/meta/src/manager/catalog/mod.rs [hunk 4]
```diff
@@ -823,271 +791,517 @@ where
...
         let mut tables = BTreeMapTransaction::new(&mut database_core.tables);
+        let mut sources = BTreeMapTransaction::new(&mut database_core.sources);
+        let mut sinks = BTreeMapTransaction::new(&mut database_core.sinks);
+        let mut views = BTreeMapTransaction::new(&mut database_core.views);
         let mut users = BTreeMapTransaction::new(&mut user_core.user_info);
 
-        let table = tables.remove(table_id);
-        if let Some(table) = table {
-            let (index_ids, index_table_ids): (Vec<_>, Vec<_>) = indexes
+        let mut deque = VecDeque::new();
```
Comment (kwannoel) [inline src/meta/src/manager/catalog/mod.rs:810]:
```suggestion
        // The deque holds all the relation_ids we need to drop.
        // As we traverse the relation DAG, 
        // more relation_ids will be added and popped from this.
        let mut deque = VecDeque::new();
```

**Review by kwannoel — APPROVED** [review]
LGTM, unifying it made it a lot more maintainable 👍

File: src/meta/src/stream/stream_manager.rs (hunk 2 above)
Comment (chenzl25) [reply src/meta/src/stream/stream_manager.rs:873]: It is a test method of `MockService` and so far so good.

**Review by yezizp2012 — APPROVED** [review]
LGTM!

**Pull request merged**